### PR TITLE
Add x_defs to rule go_binary.

### PR DESCRIPTION
**Pull request by linuxerwang**

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by damienmg** [comment]
test this please

File: go/def.bzl [hunk 1]
```diff
@@ -355,6 +356,9 @@ def emit_go_link_action(ctx, transitive_libs, lib, executable, cgo_deps):
       "-o", prefix + out,
   ]
 
+  if x_defs:
+    link_cmd += [(' -X %s ' % d) for d in x_defs]
```
Comment (damienmg) [inline go/def.bzl]: Don't you need to quote the x_defs?

**Comment by damienmg** [comment]
Just one small comment.

Also don't you have a test for it? I think reproducing the example and testing for build.time is good enough.

**Comment by linuxerwang** [comment]
Question: so I should add x_defs to go_test, and test it instead of the go_binary target, right?

**Comment by damienmg** [comment]
Yes sounds reasonable.

**Comment by linuxerwang** [comment]
It doesn't work. Something is wrong. I did see a "go tool link" call with the -X flag, but the test could not get it. Code uploaded. Please help.

**Comment by damienmg** [comment]
Let see the error from Jenkins:
retest this please

File: examples/bin/BUILD [hunk 2]
```diff
@@ -11,4 +11,7 @@ go_binary(
         "//examples/lib:go_default_library",
         "//examples/vendor/github_com/user/vendored:go_default_library",
     ],
+    x_defs = [
+        "main.buildTime=2016/05/19 09:10am",
```
Comment (damienmg) [inline examples/bin/BUILD]: this should be
```python
x_defs = {
   "main.buildTime": "2016/05/19 09:10am",
},
```

Comment (linuxerwang) [inline examples/bin/BUILD]: Sorry, my bad. Fixed.

**Comment by damienmg** [comment]
retest this please

**Comment by yugui** [comment]
The test is failing because the go importpath of `//examples/lib:go_default_library` in the generated archive is `github.com/bazelbuild/rules_go/examples/lib/go_default_library` but not `github.com/bazelbuild/rules_go/examples/lib`.
This is caused by that `emit_go_link_action` places a symlink to `//examples/lib:go_default_library` at `github.com/bazelbuild/rules_go/examples/lib/go_default_library.a` in a virtual go source tree.

I'm trying to fix the importpath in another branch. Your test should pass if you rebase your work onto my fix once I complete it.

Could you wait for the fix?  I'll work on it tomorrow or early next week.

**Comment by linuxerwang** [comment]
Thank you, Yugui. It's not an urgent feature for me, so please take your time.

**Comment by yugui** [comment]
@linuxerwang We have started a review of #28.
Once it is merged, your test should pass with the following steps.

1. Rebase your change onto #28.
2. Move the tests you added to `lib_x_test.go` from `lib_test.go`.
   
    Since `library = ":go_default_library"` lets Bazel recompile all the sources together with `lib_test`, their importpath are still `github.com/bazelbuild/rules_go/examples/lib/lib_test`.
    So you need to use `:go_default_library` as `deps` so that you can access to the string variable as `examples/lib.buildTime`

File: examples/lib/lib_test.go [hunk 3]
```diff
@@ -28,3 +28,9 @@ func TestMeaning(t *testing.T) {
 func TestAddTwoNumbers(t *testing.T) {
 	AddTwoNumbers()
 }
+
+func TestBuildTime(t *testing.T) {
+	if buildTime == "" {
```
Comment (yugui) [inline examples/lib/lib_test.go]: It is better if we can keep this test passing with just "go test" command.
So could you accept the expected value as a flag?

```go
var (
        buildTimeWant = flag.String("lib_test.buildtime", "", "expected value in TestBuildTime")
)
...
func TestBuildTime(t *testing.T) {
        if got, want := buildTime, *buildTimeWant; got != want {
               t.Errorf("buildTime = %q; want %q", got, want)
        }
}
```

You can specify the flag with `args` attribute of `go_test` rule.

**Comment by linuxerwang** [comment]
Sounds good to me. I'll rebase once #28 is merged. Thank you!

**Comment by yugui** [comment]
@linuxerwang #28 has been merged.

**Comment by linuxerwang** [comment]
Thanks. Done. I have to blaze test with:

bazel test --test_arg='--lib_test.buildtime=05/19/2016 10:30am' examples/lib:lib_external_test

Is it what you expected?

Not sure about the CLA issue. Could be because I rebased it on your change?

**Comment by yugui** [comment]
@linuxerwang 
No, you don't have to explicitly specify the flag every time.  You can write
```bzl
 go_binary(
    name = "bin",
    ...,
    x_defs = {
        "main.buildTime": "2016/05/19 09:10am",
    },
    args = [
        "-lib_test.buildtime",
        "'05/19/2016 10:30am'",
    ],
)
```

@damienmg Do you know about the CLA issue?

**Comment by damienmg** [comment]
I see the conflict resolver is not happy so I am guessing that confuse the googlebot for the CLA.

I think rebasing on top of master would be preferable.

**Comment by damienmg** [comment]
jenkins: test this please.

**Comment by linuxerwang** [comment]
> No, you don't have to explicitly specify the flag every time. You can write

Nice, thanks!

> I see the conflict resolver is not happy so I am guessing that confuse the googlebot for the CLA.

I cleaned up the commits, hope it will get resolved.

**Comment by yugui** [comment]
jenkins: retest this please

**Comment by yugui** [comment]
LGTM except the CLA:no.

I think CLA think is not working as expected because of this in the commit c5d2ccd7f61b9fc2f27b399899c0d1650d708ab4:

> Author: Zhong Wang <linuxerwang@gmail.com>
> Commit: Zhong Wang <zhwang@jingoal.com>

I guess `zhwang@jingoal.com` is not recognized as `@linuxerwang` by github or we need to make the author and the committer consistent?

**Comment by damienmg** [comment]
Oh I missed that. The cla bot needs to know about all the mail you might be
using. Two solution: re-sign the cla with the new mail or amend the commit
to change the author.

**Comment by yugui** [comment]
OK, Jenkins: retest this please

**Pull request merged**